source/language/gates.rst

```diff
@@ -164,23 +283,140 @@ of the gate definition.
 
-To avoid infinite recursion, gates must be declared before use and
+.. code-block:: c
+  gate QFT256 : qubit[8] q {
+        uint n = 8;
+        for uini j in [0 : n-1] {
+          h q[j];
+          for uint k in [j+1 : n-1]
+           ctrl @ Rz(pi / 2**(k-j+1)) q[j], q[k];
+          }
+        for uint j in [0:n-1]
+          if (j != n-1-j)
+            swap q[j], q[n-1-j]
+  }
+
+Classical storage types and parameters in a ``gate`` body are treated as being immutable,
+and cannot be assigned to more than once. (For loop induction variables are treated as being constant within
+the scope of any single iteration of the ``for`` loop, and can only be modified by the logic of the loop
+itself.) External functions, ``reset`` operations, ``measure`` operations (or other operations with potentially
+random outcomes), cannot be involved in the program logic of a ``gate`` body. This ensures that an invocation of
+a user-defined ``gate``, corresponds to a definite finite sequence of built-in unitary gates.
+
+An empty ``body`` corresponds to the identity gate.
+
+Gates must be declared before use, and 
 cannot call themselves. The statement ``name(params) qargs;`` applies the gate,
-and the variable parameters ``params`` can have any type that can promote to ``angle`` type.
+and the variable parameters ``params`` must have the appropriate type (or be expressions which can be implicitly 
+cast to the appropriate type).
+
+The quantum operands of a ``gate`` invocation must have the appropriate types, to the declaration of the ``gate``.
+There is one exception to this type-agreement condition: if a ``gate`` has one or more operands of type ``qubit``,
+the gate may instead act on some qubit register(s) *of identical size* for one or all of those operands.
+For example, using a 'short' ``gate`` declaration (all of whose operands are individual qubits):
+the quantum circuit given by
+
+.. code-block:: c
+
+   gate g qb0, qb1, qb2, qb3
+   {
+     // body
+   }
+   qubit qr0[1];
+   qubit qr1[2];
+   qubit qr2[3];
+   qubit qr3[2];
+   g qr0[0], qr1, qr2[0], qr3; // ok
+   g qr0[0], qr2, qr1[0], qr3; // error! qr2 and qr3 differ in size
+
+has a second-to-last line that means
+
+.. code-block:: c
+
+   // FIXME: insert translation of algorithmic block from TeX source.
+
+   for j in [0:1] do
+       g qr0[0],qr1[j],qr2[0],qr3[j];
+
+We provide this so that user-defined gates can be applied in parallel
+like the built-in gates. This functionality extends also to any ``gate`` declared with the
+'general' form, so long as all operands which are given an explicit size in the declaration,
+are provided with arguments of the corresponding type.
+
+.. code-block:: c
+
+   gate g : qubit qb0, qubit qb1, qubit[5] qreg
+   {
+     // body
+   }
+   qubit a[2];
+   qubit b[2];
+   qubit c[5];
+   qubit d[10];
+   g a, b, c; // ok: performs "g a[0], b[0], c; g a[1], b[1], c;"
+   g a, b, d; // error! third operand expects a register of size 5, not 10
```

Comment:
This entire section can be replaced with a reference to the broadcasting section above.